Implement JIT dispatch and compile everything to return 3

Leonardo2718 opened the pull request:
This changeset applies two important changes:

+ adds basic mechanism for dispatching the JIT and JITed code
+ implements simple IL generation that compiles everything into "return 3"

@aviansie-ben please review.

@FatalDestiny  and @TannerHagel FYI.

Leonardo2718 commented:
The changes applied here are only the bare minimum. Follow-up issues should be opened to:

- [ ] cache JIT function entry points to avoid unnecessary recompilation
- [ ] add a mechanism to disable the JIT (or just prevent JIT compilation)
- [ ] implement compilation (so functions are only compiled after they've been called a number of times)

aviansie-ben commented:
Is there any particularly compelling reason to build our JIT sources as a separate static library rather than just including them as part of our version of libwabt directly?

Leonardo2718 commented:
Not really, other than convention.

It can make it easier down the line to support "swapping" different (API compatible) versions of the interpreter and JIT without having to recompile both. It also makes it easier to migrate to a model where the JIT is dynamically linked.

aviansie-ben commented on CMakeLists.txt:
```diff
@@ -295,7 +296,7 @@ if (NOT EMSCRIPTEN)
     add_executable(${name} ${ARGV})
     add_dependencies(everything ${name})
-    target_link_libraries(${name} libwabt)
+    target_link_libraries(${name} libwabt wabtjit)
```
Please don't manually link additional libraries directly on the tool executables, since those dependencies will not propagate to external projects linking to WABT through its CMake target. Instead, set wabtjit as an interface library for libwabt by adding `PUBLIC` to your `target_link_libraries` call, as I did for jitbuilder in 23514569ed54ee1854578a5f58e30148387df1e4.

Leonardo2718 replied:
Shoot. Thought I'd fixed that. Fixing now.

aviansie-ben commented on src/jit/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+    wabtjit.cpp
+)
+
+target_link_libraries(wabtjit jitbuilder)
```
If we're going to link jitbuilder here, this should probably be marked as `PUBLIC` and the `target_link_libraries` call for jitbuilder being linked into libwabt should be removed.

This also causes a regression on #6, since jitbuilder include files are not set as system include files in the context of JIT source files. Please adapt the code that solved that issue (see 4bee644c466fdc8e992de3d62e63e5b4358fdf5b) to work on the wabtjit target.

aviansie-ben commented on src/jit/function-builder.cpp:
```diff
@@ -0,0 +1,104 @@
+
+    DefineReturnType(types->toIlType<std::underlying_type<wabt::interp::Result>::type>());
+//    setUseBytecodeBuilders();
+//    AllLocalsHaveBeenDefined();
```
I'm not sure what to make of these commented out calls. If this was intentional, please leave a TODO documenting why they've been commented out to make sure they get re-added when necessary.

Leonardo2718 replied:
Those are calls I originally added because we will (probably) eventually need them, just not at the moment.

I've removed them for now as they really should be added as part of the work that needs them.

Leonardo2718 commented:
@aviansie-ben 

You're absolutely right about the code style. I was actually trying to match what WABT uses, but clearly, I wasn't paying enough attention 😛 . I've fixed the style according to your suggestions. I've also updated the file names to match the style used by WABT.

Let me know if I've missed anything 🙂 .

aviansie-ben commented on src/jit/function-builder.cc:
```diff
@@ -0,0 +1,102 @@
+      offset_{offset},
+      valueType_{types->LookupUnion("Value")},
+      pValueType_{types->PointerTo(types->LookupUnion("Value"))}
+{
```
This open brace should go at the end of the previous line.

aviansie-ben commented on src/jit/function-builder.cc:
```diff
@@ -0,0 +1,102 @@
+      thread_{thread},
+      offset_{offset},
+      valueType_{types->LookupUnion("Value")},
+      pValueType_{types->PointerTo(types->LookupUnion("Value"))}
```
WABT seems to use exclusively parentheses for initializing objects, so we should probably stick to that convention where possible.

aviansie-ben commented on src/jit/wabtjit.cc:
```diff
@@ -0,0 +1,33 @@
+  FunctionBuilder builder{thread, offset, &types};
+  uint8_t* function = nullptr;
+
+  if ( compileMethodBuilder(&builder, &function) == 0 ) {
```
There shouldn't be spaces around the condition here.

aviansie-ben commented on CMakeLists.txt:
```diff
@@ -272,7 +272,6 @@ function(add_omr)
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -w")
 
   add_subdirectory(third_party/omr EXCLUDE_FROM_ALL)
-  target_link_libraries(libwabt PUBLIC jitbuilder)
```
You removed the actual linking here, but you forgot to remove the part below that marks the include directories as system. Since we're inheriting the include directories of wabtjit (which already marks them as system), this shouldn't be needed anymore.

aviansie-ben commented on src/jit/function-builder.cc:
```diff
@@ -0,0 +1,102 @@
+  DefineFile(__FILE__);
+  DefineName("WASM_Function");
+
+
```
There's an extra empty line here that can be removed.

Leonardo2718 commented:
@aviansie-ben fixed.
